src/sage/categories/modules_with_basis.py

```diff
@@ -2062,17 +2063,25 @@ def trailing_term(self, *args, **kwds):
             """
             return self.parent().term(*self.trailing_item(*args, **kwds))
 
-        def map_coefficients(self, f):
+        def map_coefficients(self, f, new_base_ring=None):
             """
-            Mapping a function on coefficients.
+            Return the element obtained by applying ``f`` to the non-zero
+            coefficients of ``self``.
+
+            If ``f`` is a :class:`sage.categories.map.Map`, then the resulting
+            polynomial will be defined over the codomain of ``f``. Otherwise, the
+            resulting polynomial will be over the same ring as ``self``. Set
+            ``new_base_ring`` to override this behaviour.
+
+            An error is raised if the coefficients are not in the new base ring.
```

Comment:
As written, I don't think we want this error to be raised if the callable returns elements in, say `ZZ` but `new_base_ring=GF(2)`. We want to automatically convert the elements for consistency:
```
sage: R.<x,y> = ZZ[]
sage: x.map_coefficients(lambda p: 2*p, GF(2))
0
sage: x.map_coefficients(lambda p: GF(2)(p), QQ)
x
```
I think what you meant to say is if the coefficients must be able to be converted into `new_base_ring`.

Additionally, if `f` is a map, then `new_base_ring` should still be the ultimate target ring for consistency:
```
sage: R.<a> = QQ[]
sage: f = R.hom([a^2])
sage: S.<x,y> = R[]
sage: (a*x).map_coefficients(f)
a^2*x
sage: (a*x).map_coefficients(f, R['q'])
a^2*x
sage: _.parent()
Multivariate Polynomial Ring in x, y over Univariate Polynomial Ring in q over Univariate Polynomial Ring in a over Rational Field
```

Reply:
Do I understand correctly that you are asking me to adapt the documentation, the code is doing what it should, right?

Reply:
Yes, that's correct.